### ISSUE TITLE
no-shadow false positive on assigning a class to a var of the same name

### ISSUE THREAD

**Issue by MarkKahn** [eslint/eslint#10374]
[minimal repro](https://eslint.org/demo/#eyJ0ZXh0IjoibGV0IEZvbztcblxuRm9vID0gY2xhc3MgRm9vIHt9Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo2LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7ImNvbnN0cnVjdG9yLXN1cGVyIjoyLCJuby1jYXNlLWRlY2xhcmF0aW9ucyI6Miwibm8tY2xhc3MtYXNzaWduIjoyLCJuby1jb21wYXJlLW5lZy16ZXJvIjoyLCJuby1jb25kLWFzc2lnbiI6Miwibm8tY29uc29sZSI6Miwibm8tY29uc3QtYXNzaWduIjoyLCJuby1jb25zdGFudC1jb25kaXRpb24iOjIsIm5vLWNvbnRyb2wtcmVnZXgiOjIsIm5vLWRlYnVnZ2VyIjoyLCJuby1kZWxldGUtdmFyIjoyLCJuby1kdXBlLWFyZ3MiOjIsIm5vLWR1cGUtY2xhc3MtbWVtYmVycyI6Miwibm8tZHVwZS1rZXlzIjoyLCJuby1kdXBsaWNhdGUtY2FzZSI6Miwibm8tZW1wdHktY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1lbXB0eS1wYXR0ZXJuIjoyLCJuby1lbXB0eSI6Miwibm8tZXgtYXNzaWduIjoyLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOjIsIm5vLWV4dHJhLXNlbWkiOjIsIm5vLWZhbGx0aHJvdWdoIjoyLCJuby1mdW5jLWFzc2lnbiI6Miwibm8tZ2xvYmFsLWFzc2lnbiI6Miwibm8taW5uZXItZGVjbGFyYXRpb25zIjoyLCJuby1pbnZhbGlkLXJlZ2V4cCI6Miwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOjIsIm5vLW1peGVkLXNwYWNlcy1hbmQtdGFicyI6Miwibm8tbmV3LXN5bWJvbCI6Miwibm8tb2JqLWNhbGxzIjoyLCJuby1vY3RhbCI6Miwibm8tcmVkZWNsYXJlIjoyLCJuby1yZWdleC1zcGFjZXMiOjIsIm5vLXNlbGYtYXNzaWduIjoyLCJuby1zcGFyc2UtYXJyYXlzIjoyLCJuby10aGlzLWJlZm9yZS1zdXBlciI6Miwibm8tdW5kZWYiOjIsIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjoyLCJuby11bnJlYWNoYWJsZSI6Miwibm8tdW5zYWZlLWZpbmFsbHkiOjIsIm5vLXVuc2FmZS1uZWdhdGlvbiI6Miwibm8tdW51c2VkLWxhYmVscyI6Miwibm8tdXNlbGVzcy1lc2NhcGUiOjIsInJlcXVpcmUteWllbGQiOjIsInVzZS1pc25hbiI6MiwidmFsaWQtdHlwZW9mIjoyLCJuby1zaGFkb3ciOjJ9LCJlbnYiOnt9fX0=)

```
let Foo;

Foo = class Foo {} // 3:13 - 'Foo' is already declared in the upper scope. (no-shadow)
```

`Foo` in this case refers to the same thing, so there's no issue and I believe this is a false positive

**Comment by kaicataldo** [comment]
For context this also warns with `3:16 - 'foo' is already declared in the upper scope. (no-shadow)`:

```
let foo;

foo = function foo() {}
```

**Comment by platinumazure** [comment]
For both function names and class names, I believe those are in a child scope (between the containing scope and the class's/function's inner scope), so they are shadowing.

In both cases, you can use anonymous functions/classes:

```js
let foo = function () {};

let bar = class {};
```

I could potentially support adding an option to no-shadow which would allow the shadowing as long as the lower scope is function/class name, the function/class expression is the RHS of an assignment, and the function/class name is also being used on the LHS of the assignment.

**Comment by MarkKahn** [comment]
They might be in different scopes, but only by the strict definition of the spec.  They're not in practice.

The rule is meant to protect against this:

```js
let a = 42;
let b = a;
function() {
  let a = 43;
  // a !== b
}
```

But in the scenario posted:

```js
let a = function b() {
  // a === b
}
```

In other words there is no practical way (save a bug in IE6) that this is actually a different scope.

The problem with anonymous classes/functions is that you lose information like `this.constructor.name`, primarily useful for debugging

**Comment by MarkKahn** [comment]
Also note that this is fine:

    let a = function a() {
    }
